File: engine/src/elements/Stack.ts
import Element from '../Element'
import BoxConstraints from '../BoxConstraints'
import Size from '../Size'
import Position from '../Position'

interface StackProps {
	children: Array<Element<any>>
}

class Stack extends Element<StackProps> {
	layout(constraints: BoxConstraints) {
		for (let child of this.props.children) {
			child.size = this.document.layoutElement(child, constraints.makeLoose())
			child.relPosition = Position.origin
		}
		return new Size({
			width: constraints.maxWidth,
			height: constraints.maxHeight
		})
	}

	paint() {
		for (let child of this.props.children) {
			this.document.paintElement(child)
		}
	}
}

export default Stack

File: core/libs/ui/src/elements/stack.cpp
#include "elements/stack.hpp"

namespace aardvark {

Size StackElement::layout(BoxConstraints constraints) {
    auto max_width = 0.0f;
    auto max_height = 0.0f;
    auto floating_children = std::vector<Element*>();
    for (auto& child : children) {
        auto stack_child = dynamic_cast<StackChildElement*>(child.get());
        if (stack_child != nullptr && stack_child->floating) {
            floating_children.push_back(child.get());
            continue;
        }
        child->size = document->layout_element(
            child.get(),
            loosen_constraints ? constraints.make_loose() : constraints);
        child->rel_position = Position{0, 0};
        max_width = fmax(max_width, child->size.width);
        max_height = fmax(max_height, child->size.height);
    }
    auto floating_constraints = BoxConstraints{
        loosen_constraints ? 0 : constraints.min_width,   // min_width
        max_width,                                        // max_width
        loosen_constraints ? 0 : constraints.min_height,  // min_height
        max_height                                        // max_height
    };
    for (auto child : floating_children) {
        child->size = document->layout_element(child, floating_constraints);
        child->rel_position = Position{0, 0};
    }
    return Size{max_width, max_height};
}

}  // namespace aardvark
